File: HyLang/Parser/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using HyLang.Parser.AST;

namespace HyLang.Parser;

public class Parser
{
    private int _position = 0;
    private List<Token> _tokens;

    public Parser(List<Token> tokens) => _tokens = tokens;

    public List<Node> ParseCode()
    {
        var root = new List<Node>();
        while (_position < _tokens.Count)
        {
            var expression = Expression();
            if (!MatchToken(TokenType.End))
            {
                throw new Exception();
            }
            root.Add(expression);
        }

        return root;
    }

    private Node Expression()
    {
        return Additive();
    }

    private Node Additive(){
        var result = Multiplicative();
                while (true)
        {
            if( MatchToken(TokenType.Plus)){
                result = new BinaryExpressionNode(result ,"+" , Multiplicative() );
            }
            else if( MatchToken(TokenType.Minus)){
                result = new BinaryExpressionNode(result ,"-" , Multiplicative() );
            }
            else{
                break;
            }
        }
        return result;
    }
    private Node Multiplicative(){
        var result = Unary();
        while (true)
        {
            if( MatchToken(TokenType.Multiply)){
                result = new BinaryExpressionNode(result ,"*" , Unary() );
            }
            else if( MatchToken(TokenType.Divide)){
                result = new BinaryExpressionNode(result ,"/" , Unary() );
            }
            else{
                break;
            }
        }
        return result;
    }
    private Node Unary(){

        var result = Primary();
        return result;
    }
    private Node Primary(){
        var token = GetCurrentToken();
        if(MatchToken(TokenType.Number)){
            return new NumberExpressionNode(token.Text);
        }
        else if(MatchToken(TokenType.LeftParenthesis)){
            var result = Expression();
            RequireToken(TokenType.RightParenthesis);
            return result;
        }
        throw new Exception("Symbol error");
    }


    private Token GetCurrentToken(int relativePos = 0)
    {
        var token = _tokens[_position+relativePos];
        return token;
    }
    private void RequireToken(params TokenType[] types){
        if(!MatchToken(types))
            throw new Exception($"Require {types[0]}");
    }
    private bool MatchToken(params TokenType[] types)
    
    {
        var token = GetCurrentToken();
        if (types.Contains(token.Type))
        {
            _position++;
            return true;
        }

        return false;
    }
}

File: src/compiler/parser/parser.cpp
#include "parser.h"

#include "../ast/nodes/binaryExpNode.h"
#include "../ast/nodes/integerExpNode.h"
#include "../ast/nodes/unaryExpNode.h"

std::unique_ptr<Node> Parser::primary()
{
    auto token = tokens[index++];
    int number = 0;
    std::unique_ptr<Node> expr;
    switch (token.type)
    {
    case TokenType::NUMBER:
        number = std::atoi(chars.substr(token.start, token.end - token.start).c_str());
        return std::make_unique<IntegerExpNode>(number);
    case TokenType::LEFT_PARENTHESIS:
        expr = expression();
        require_token(TokenType::RIGHT_PARENTHESIS);
        return expr;
    // default: break;//throw std::runtime_error("Invalid token");
    }
}

std::unique_ptr<Node> Parser::unary()
{
    auto token = tokens[index];
    std::unique_ptr<Node> res;
    switch (token.type)
    {
    case TokenType::MINUS:
        index++;
        res = primary();
        return std::make_unique<UnaryExpNode>(std::move(res),UnaryOperation::Negate);
    default: return primary();
    }
}

std::unique_ptr<Node> Parser::multiplicative()
{
    std::unique_ptr<Node> res = unary();
    std::unique_ptr<Node> un;
    while (true) {
        auto token = tokens[index];
        switch (token.type)
        {
            case TokenType::STAR:
                index++;
                un = unary();
                res = std::make_unique<BinaryExpNode>(std::move(res), std::move(un), BinaryOperation::Multiply);
                break;
            case TokenType::SLASH:
                index++;
                un = unary();
                res = std::make_unique<BinaryExpNode>(std::move(res), std::move(un), BinaryOperation::Divide);
                break;
            default: return res;
        }
    }

}

std::unique_ptr<Node> Parser::additive()
{
    std::unique_ptr<Node> res = multiplicative();
    std::unique_ptr<Node> mul;
    while (true) {
        auto token = tokens[index];
        switch (token.type)
        {
            case TokenType::PLUS:
                index++;
                mul = multiplicative();
                res = std::make_unique<BinaryExpNode>(std::move(res), std::move(mul), BinaryOperation::Add);
                break;
            case TokenType::MINUS:
                index++;
                mul = multiplicative();
                res = std::make_unique<BinaryExpNode>(std::move(res), std::move(mul), BinaryOperation::Subtract);
                break;
            default: return res;
        }
    }
}

std::unique_ptr<Node> Parser::conditional()
{
    return additive();
}

std::unique_ptr<Node> Parser::expression()
{
    return conditional();
}

void Parser::require_token(TokenType type)
{
    auto token = tokens[index];
    if(token.type == type)
    {
        index++;
        return;
    }

    //throw
}

Parser::Parser(std::vector<Token>& tokens, std::string& chars): tokens(tokens), chars(chars), index(0)
{
}

std::unique_ptr<Node> Parser::parse()
{
    return expression();
}
